pkg/kubelet/cm/cpumanager/topology_hints.go

```diff
@@ -0,0 +1,149 @@
+	// The 'none' policy does not generate topology hints.
+	if m.policy.Name() == string(PolicyNone) {
+		return nil
+	}
+
+	// For all other policies, if there are no CPU resources requested for this
+	// container, we do not generate any topology hints.
+	if _, ok := container.Resources.Requests[v1.ResourceCPU]; !ok {
+		return nil
+	}
+
+	// Otherwise, attempt to generate TopologyHints for the CPUManager.
+	// For now, this implementation assumes the 'static' CPUManager policy.
+	// TODO: Generalize this so that its applicable to future CPUManager polices.
+
+	// Get a count of how many guaranteed CPUs have been requested.
+	requested := m.policy.(*staticPolicy).guaranteedCPUs(&pod, &container)
+
+	// If there are no guaranteed CPUs being requested, we do not generate
+	// any topology hints. This can happen, for example, because init
+	// containers don't have to have guaranteed CPUs in order for the pod
+	// to still be in the Guaranteed QOS tier.
+	if requested == 0 {
+		return nil
+	}
+
+	// Get a list of available CPUs.
+	available := m.policy.(*staticPolicy).assignableCPUs(m.state)
+
+	// Generate hints.
+	cpuHints := m.generateCPUTopologyHints(available, requested)
+	klog.Infof("[cpumanager] TopologyHints generated for pod '%v', container '%v': %v", pod.Name, container.Name, cpuHints)
+
+	return cpuHints
+}
+
+// iterateSocketCombinations iterates through all combinations of 'sockets' of any size.
+func (m *manager) iterateSocketCombinations(sockets cpuset.CPUSet, callback func(cpuset.CPUSet)) {
+	// Internal helper function to accumulate the combination before calling the callback.
+	var iterate func(sockets, accum []int, amount int)
+	iterate = func(sockets, accum []int, amount int) {
+		if len(accum) == amount {
+			callback(cpuset.NewCPUSet(accum...))
+			return
+		}
+		for i := range sockets {
+			iterate(sockets[i+1:], append(accum, sockets[i]), amount)
+		}
+	}
+
+	for i := 1; i <= sockets.Size(); i++ {
+		iterate(sockets.ToSliceNoSort(), []int{}, i)
+	}
+}
+
+// generateCPUtopologyHints generates a set of TopologyHints given the set of
+// available CPUs and the number of CPUs being requested.
+//
+// It follows the convention of marking all hints that have the same number of
+// bits set as the narrowest matching SocketAffinity with 'Preferred: true', and
+// marking all others with 'Preferred: false'.
+func (m *manager) generateCPUTopologyHints(availableCPUs cpuset.CPUSet, request int) []topologymanager.TopologyHint {
+	// Discover topology in order to establish the number
+	// of available CPUs per socket.
+	topo, err := topology.Discover(m.machineInfo)
+	if err != nil {
+		klog.Warningf("[cpu manager] Error discovering topology for TopologyHint generation")
+		return nil
+	}
+
+	// Initialize minAffinity to a full affinity mask.
+	minAffinity, _ := socketmask.NewSocketMask()
+	minAffinity.Fill()
+
+	// Iterate through all combinations of sockets
```

Comment:
**NOTE:** It looks like this algorithm is `O(n!)` in number of sockets. 8 sockets already yields ~40k combinations to check. Nowadays IIUC >8 socket systems are pretty exotic, but maybe some kind of guard would help prevent this from crashing/stalling the kubelet.

Reply:
Adding a couple tests with high socket counts (4 and 8) would also help.

Reply:
Yes. The builtin assumption is that the number of sockets is relatively low, so this doesn't explode. I am adding some tests now for 4 and 8 socket system.

In theory, (independent of the CPUManager) we should be generating one hint for every possible socket combination that can satisfy the incoming request (otherwise we miss aligning with an exotic combination from another HintProvider). As such, I don't see a way around enumerating all socket combinations in the worst case (i.e. O(n!)).

I'm all for putting a guard in to ensure the socket count is low enough not to stall the kubelet (I think 8 is a reasonable number to choose too). Where would you recommend putting this guard in? Ideally we would have it fail the kubelet on startup (as opposed to just failing once a pod is submitted), saying we don't support the use of the TopologyManager on systems with > 8 Sockets.

Reply:
That seems reasonable to me too, but may want to get a Kubelet maintainer to chime in on their preference if they are available to comment.

/cc @derekwaynecarr

Reply:
Also, it's not quite O(n!). Its actually:

`sum(n choose k) for k=1...numSockets`

which is much less than n! for large n.

Compare:
```
2 sockets: (n! = 2), (sum(n choose k) = 3)
4 sockets: (n! = 24), (sum(n choose k) = 15)
8 sockets: (n! = 40320),  (sum(n choose k) = 255)
16 sockets: (n! = 20922789888000), (sum(n choose k) = 65535)
```

Reply:
Tests added.

Reply:
Thanks for clarifying the complexity Kevin. Let's follow up with the discussion of guarding topology manager startup on supported number of sockets in an issue and/or upcoming sig-node meeting.

Reply:
Added to the sig-node meeting for 13-Aug-2019

Reply:
Following up, plan discussed for alpha in this release cycle is to fail-fast for large number of socket in the topology manager initialization routine. That will be done in a subsequent PR.